### DISCUSSION TITLE
Row/column numbering not from 0 when using dzsave with `--layout google` and `--centre`

### DISCUSSION THREAD

**Discussion by TimPushkin** [libvips/libvips#3766] [Q&A] [answered]
When I use dzsave with `--layout google` and `--centre` options numbering of columns and rows does not start from 0 and I have to renumber them (which is not easy when `--skip-blanks` is also used since the first tiles may be skipped and then I don't know the index of the 0th row/column). Is there a way to make libvips number rows and columns starting from 0?

**Comment by jcupitt** [comment]
Hi again, it's just like skip blanks -- the idea is that any missing tile can be replaced by `blank.png`. They are not numbered from 0 because centring will add a large blank border to the image edges in order to fit the whole image in one tile.

**Reply by TimPushkin** [reply]
I thought so, but is there a way to get the tiles numbered from 0 on each zoom level?

**Reply by jcupitt** [reply]
They should be numbered from zero, if you include the border added by `--centre`. I'm not sure I understand your question.

**Reply by jcupitt** [reply]
For example, with this landscape photo:

![nina](https://github.com/libvips/libvips/assets/580843/af24b15c-78f3-4195-9a98-4b2cc46b7fe0)

If I run:

```
$ vips dzsave nina.jpg x --layout google --centre --background 128
```

(using a grey background to make it easy to see)

I see this for tile `x/0/0/0.jpg`:

![0](https://github.com/libvips/libvips/assets/580843/64529a2f-18c7-4824-ad1f-5383387af375)

You really get a huge border, especially at the top. If I look in level 5 (the highest res) I see:

```
$ ls x/5/
10/ 11/ 12/ 13/ 14/ 15/ 16/ 17/ 18/ 19/ 20/ 21/ 22/ 23/ 8/  9/  
$ ls x/5/8/
10.jpg  13.jpg  16.jpg  19.jpg  22.jpg  25.jpg  4.jpg   7.jpg   
11.jpg  14.jpg  17.jpg  20.jpg  23.jpg  26.jpg  5.jpg   8.jpg   
12.jpg  15.jpg  18.jpg  21.jpg  24.jpg  27.jpg  6.jpg   9.jpg   
```

So tile (4, 8) is the first one:

![4](https://github.com/libvips/libvips/assets/580843/87b2aa6f-539b-49e1-92e3-7a4f20fc92ef)

And it's the top-left of the image. Everything up and left of that is background.

**Reply by TimPushkin** [reply]
UPD: sorry, I wrote this before seeing you message above, reading it now.

They are actually not or maybe I still don't understand something... For example, I have this image (a rectangle with transparent surroundings):
![test](https://github.com/libvips/libvips/assets/71890197/0d538d4d-0f12-4fab-83fb-03054c8fd0d3)

When I do `vips dzsave "test.png" "test" --layout google --tile-size 64 --depth onetile --overlap 0 --background 0 --skip-blanks -1 --centre --suffix .png`, as I understand `--skip-blanks -1` makes vip not delete blanks, so the numbering should start from 0, if I understand you right. But it does not, e.g. on zoom level 2 I get rows 1, 2, and on level 3 I get 2, 3, 4, 5. Same with columns, but it'll require another example image to demonstrate.

What I actually want to get is for my images to be centered, but for the padding to be added only on lower-resolution levels where it is needed, but not on the higher resolution levels, where the image spans from start to end by itself. And at the same time I want the numbering of higher resolution levels to start from zero. Not sure if my explanation is clear, but with the example image above if the rows were numbered 0, 1 on level 2 and 0, 1, 2, 3 on level 3 this would be exactly what I need since vips seems to skip the padding tiles by itself.

**Reply by jcupitt** [reply]
In google centre mode it never outputs tiles which it knows must be equal to the background. 

Skip blanks is an extra test applied to real image tiles to check for areas of background, it's useful for microscopy images like this:

https://openslide.org/demo/

(click on one of the slides on the left)

Where really large parts of the image can just be flat white.

**Reply by TimPushkin** [reply]
So, in these terms, I want the tiles that cover the real image to be numbered from 0, but it seems like this is not what vips can easily do, so I'll need to renumber them myself, manually ensuring the top left corner of my image is not removed as blank.

But thanks again for the explanation!

**Reply by jcupitt** [reply, marked as answer]
Your levels may not line up if you renumber, of course, so you need to be a bit cautious. It depends on your tile numbering scheme.

If you are targetting one of the (many!) javascript map viewers, they usually have a function you can override for tile fetching, eg. `getTile(level, x, y)`. It might be easier to fix any numbering differences in that.
